app/models/concerns/alegre_v2.rb

```diff
@@ -274,15 +308,72 @@ def get_similar_items_v2(project_media, field)
       end
     end
 
+    def get_similar_items_v2_async(project_media, field)
+      type = get_type(project_media)
+      if !Bot::Alegre.should_get_similar_items_of_type?('master', project_media.team_id) || !Bot::Alegre.should_get_similar_items_of_type?(type, project_media.team_id)
+        return false
+      else
+        get_suggested_items_async(project_media, field)
+        get_confirmed_items_async(project_media, field)
+        return true
+      end
+    end
+
+    def get_required_keys(project_media, field)
+      {
+        confirmed_results: "alegre:async_results:#{project_media.id}_#{field}_true",
+        suggested_or_confirmed_results: "alegre:async_results:#{project_media.id}_#{field}_false"
+      }
+    end
+
+    def get_cached_data(required_keys)
+      redis = Redis.new(REDIS_CONFIG)
+      Hash[required_keys.collect{|k,v| [k, (Hash[JSON.parse(redis.get(v)).collect{|kk,vv| [kk.to_i, vv]}] rescue [])]}]
+    end
+
+    def get_similar_items_v2_callback(project_media, field)
+      type = get_type(project_media)
+      if !Bot::Alegre.should_get_similar_items_of_type?('master', project_media.team_id) || !Bot::Alegre.should_get_similar_items_of_type?(type, project_media.team_id)
+        return {}
+      else
+        cached_data = get_cached_data(get_required_keys(project_media, field))
+        if !cached_data.values.include?(nil)
+          suggested_or_confirmed = cached_data[:suggested_or_confirmed_results]
+          confirmed = cached_data[:confirmed_results]
+          Bot::Alegre.merge_suggested_and_confirmed(suggested_or_confirmed, confirmed, project_media)
+        end
+      end
+    end
+
     def relate_project_media(project_media, field=nil)
       self.add_relationships(project_media, self.get_similar_items_v2(project_media, field)) unless project_media.is_blank?
     end
 
+    def relate_project_media_async(project_media, field=nil)
+      self.get_similar_items_v2_async(project_media, field) unless project_media.is_blank?
+    end
+
+    def relate_project_media_callback(project_media, field=nil)
+      self.add_relationships(project_media, self.get_similar_items_v2_callback(project_media, field)) unless project_media.is_blank?
+    end
+
     def get_items_with_similar_media_v2(media_url, threshold, team_ids, type)
       alegre_path = ['audio', 'image'].include?(type) ? self.sync_path_for_type(type) : "/#{type}/similarity/search/"
       # FIXME: Stop using this method from v1 once all media types are supported by v2
       # FIXME: Alegre crashes if `media_url` was already requested before, this is why I append a hash
       self.get_items_with_similar_media("#{media_url}?hash=#{SecureRandom.hex}", threshold, team_ids, alegre_path)
     end
```

Comment:
@DGaffney, can this method still work as sync for now? I'm thinking about two use cases beyond regular similarity matching that would require further changes to the workflow: tipline search and feed search.